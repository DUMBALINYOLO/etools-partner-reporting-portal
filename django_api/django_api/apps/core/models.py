from __future__ import unicode_literals

import logging
from datetime import datetime, timedelta
from decimal import Decimal

import pycountry
from django.conf import settings
from django.contrib.gis.db import models
from django.core.validators import (
    MinValueValidator,
    MaxValueValidator
)
from django.db.models import Q
from django.utils.encoding import python_2_unicode_compatible
from django.utils.functional import cached_property

from model_utils.models import TimeStampedModel
from mptt.models import MPTTModel, TreeForeignKey

from core.countries import COUNTRY_NAME_TO_ALPHA2_CODE
from .common import (
    RESPONSE_PLAN_TYPE,
    INDICATOR_REPORT_STATUS,
    OVERALL_STATUS,
    EXTERNAL_DATA_SOURCES,
    PRP_ROLE_TYPES,
)
from utils.emails import send_email_from_template
from utils.groups.wrappers import GroupWrapper


logger = logging.getLogger('locations.models')

# Define groups that UNICEF IP reporting users can belong to
try:
    PartnerAuthorizedOfficerRole = GroupWrapper(
        code='ip_authorized_officer', name='IP Authorized Officer', create_group=False
    )
    PartnerEditorRole = GroupWrapper(
        code='ip_editor', name='IP Editor', create_group=False
    )
    PartnerViewerRole = GroupWrapper(
        code='ip_viewer', name='IP Viewer', create_group=False
    )
    IMORole = GroupWrapper(code='imo', name='IMO', create_group=False)
except Exception as e:
    print("Group DB is not ready yet! - Error: %s" % e)


class TimeStampedExternalSyncModelMixin(TimeStampedModel):
    """
    A abstract class that provides external_id field that some models need since
    they might have been synced from an external system.
    """
    external_id = models.CharField(
        help_text='An ID representing this instance in an external system',
        blank=True,
        null=True,
        max_length=32
    )

    class Meta:
        abstract = True


class TimeStampedExternalSourceModel(TimeStampedExternalSyncModelMixin):
    external_source = models.TextField(choices=EXTERNAL_DATA_SOURCES, blank=True, null=True)

    class Meta:
        abstract = True
        unique_together = ('external_id', 'external_source')


class Country(TimeStampedModel):
    """
    Represents a country which has many offices and sections.
    Taken from https://github.com/unicef/etools/blob/master/EquiTrack/users/models.py
    on Sep. 14, 2017.
    """
    name = models.CharField(max_length=100)
    country_short_code = models.CharField(
        max_length=10,
        null=True,
        blank=True
    )
    long_name = models.CharField(max_length=255, null=True, blank=True)

    class Meta:
        ordering = ['name']

    def __str__(self):
        return self.name

    @property
    def details(self):
        """
        Tries to retrieve a usable country reference
        :return: pycountry Country object or None
        """
        lookup = None

        if not self.country_short_code:
            lookup = {'alpha_2': COUNTRY_NAME_TO_ALPHA2_CODE.get(self.name, None)}
        elif len(self.country_short_code) == 3:
            lookup = {'alpha_3': self.country_short_code}
        elif len(self.country_short_code) == 2:
            lookup = {'alpha_2': self.country_short_code}

        if lookup:
            try:
                return pycountry.countries.get(**lookup)
            except KeyError:
                pass


class Workspace(TimeStampedExternalSourceModel):
    """
    Workspace (previously called Workspace, also synonym was
    emergency/country) model.

    It's used for drop down menu in right top corner in the UI. Many times
    workspace is associated with only one country.
    """
    title = models.CharField(max_length=255)
    workspace_code = models.CharField(
        max_length=8,
        unique=True
    )
    countries = models.ManyToManyField(Country, related_name='workspaces')
    business_area_code = models.CharField(
        max_length=10,
        null=True, blank=True
    )
    latitude = models.DecimalField(
        null=True, blank=True,
        max_digits=8, decimal_places=5,
        validators=[
            MinValueValidator(Decimal(-90)),
            MaxValueValidator(Decimal(90))
        ]
    )
    longitude = models.DecimalField(
        null=True, blank=True,
        max_digits=8, decimal_places=5,
        validators=[
            MinValueValidator(Decimal(-180)),
            MaxValueValidator(Decimal(180))
        ]
    )
    initial_zoom = models.IntegerField(default=8)

    class Meta:
        ordering = ['title']

    def __str__(self):
        return self.title

    @property
    def locations(self):
        """
        Returns a list of locations that belong to countries associated with
        this workspace.
        """
        result = self.countries.all().values_list(
            'gateway_types__locations').distinct()
        pks = []
        [pks.extend(filter(lambda x: x is not None, part)) for part in result]
        return Location.objects.filter(pk__in=pks)

    @property
    def can_import_ocha_response_plans(self):
        return any([
            c.details for c in self.countries.all()
        ])


class PRPRole(TimeStampedExternalSourceModel):
    """
    PRPRole model present a workspace-partner level permission entity
    with cluster association.

    related models:
        account.User (ForeignKey): "user"
    """
    user = models.ForeignKey(
        'account.User', related_name="prp_roles"
    )
    role = models.CharField(max_length=32, choices=PRP_ROLE_TYPES)
    workspace = models.ForeignKey(
        'core.Workspace', related_name="prp_roles",
        null=True, blank=True
    )
    cluster = models.ForeignKey(
        'cluster.Cluster', related_name="prp_roles",
        blank=True, null=True
    )

    def __str__(self):
        return '{} - {} in Workspace {}'.format(self.user, self.role, self.workspace)

    def send_email_notification(self, created):
        template_data = {
            'user': self.user,
            'role': self,
            'id_management_url': f'{settings.FRONTEND_HOST}/id-management/',
        }
        to_email_list = [self.user.email]
        content_subtype = 'html'

        if self.role == PRP_ROLE_TYPES.ip_admin:
            subject_template_path = 'emails/ip/notify_newly_designated_ip_admin_subject.txt'
            body_template_path = 'emails/ip/notify_newly_designated_ip_admin.html'
            template_data['training_materials_url'] = '#'  # TBD
        elif self.role in {PRP_ROLE_TYPES.ip_editor, PRP_ROLE_TYPES.ip_viewer}:
            if created:
                subject_template_path = 'emails/ip/notify_on_role_assignment_subject.txt'
                body_template_path = 'emails/ip/notify_on_role_assignment.html'
                template_data['training_materials_url'] = '#'  # TBD
            else:
                subject_template_path = 'emails/ip/notify_on_role_edit_subject.txt'
                body_template_path = 'emails/ip/notify_on_role_edit.html'
                template_data['training_materials_url'] = '#'   # TBD
        elif self.role == PRP_ROLE_TYPES.cluster_imo and created:
            subject_template_path = 'emails/cluster/notify_imo_on_role_assignment_subject.txt'
            body_template_path = 'emails/cluster/notify_imo_on_role_assignment.html'
            template_data['training_materials_url'] = '#'  # TBD
        elif self.role in {PRP_ROLE_TYPES.cluster_member,
                           PRP_ROLE_TYPES.cluster_viewer,
                           PRP_ROLE_TYPES.cluster_coordinator} and created:
            subject_template_path = 'emails/cluster/notify_on_role_assignment_subject.txt'
            body_template_path = 'emails/cluster/notify_on_role_assignment.html'
            template_data['training_materials_url'] = '#'  # TBD
            template_data['system_admin_email'] = '#'  # TBD
        else:
            return False

        send_email_from_template(
            subject_template_path=subject_template_path,
            body_template_path=body_template_path,
            template_data=template_data,
            to_email_list=to_email_list,
            content_subtype=content_subtype
        )
        return True


class ResponsePlan(TimeStampedExternalSourceModel):
    """
    ResponsePlan model present response of workspace (intervention).

    related models:
        ....
    """
    title = models.CharField(max_length=255, verbose_name='Response Plan')
    plan_type = models.CharField(
        max_length=3,
        choices=RESPONSE_PLAN_TYPE,
        default=RESPONSE_PLAN_TYPE.hrp,
        verbose_name='Plan Type'
    )
    start = models.DateField(
        null=True,
        blank=True,
        verbose_name='Start date'
    )
    end = models.DateField(
        null=True,
        blank=True,
        verbose_name='End date'
    )
    workspace = models.ForeignKey(
        'core.Workspace', related_name="response_plans"
    )

    class Meta:
        unique_together = ('title', 'plan_type', 'workspace')

    def __str__(self):
        return '#{} {}'.format(self.id, self.title)

    @property
    def documents(self):
        return []  # TODO probably create file field

    @cached_property
    def all_clusters(self):
        return self.clusters.all()

    @cached_property
    def can_import_ocha_projects(self):
        """
        We need external id and source to search projects for this plan
        """
        return bool(
            self.external_id and self.external_source == EXTERNAL_DATA_SOURCES.HPC
        )

    def num_of_partners(self, clusters=None):
        from partner.models import Partner

        if not clusters:
            clusters = self.all_clusters
        return Partner.objects.filter(clusters__in=clusters).distinct().count()

    def num_of_due_overdue_indicator_reports(self,
                                             clusters=None,
                                             partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_due_overdue_indicator_reports(partner=partner)
        return count

    def num_of_non_cluster_activities(self,
                                      clusters=None,
                                      partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_non_cluster_activities(partner=partner)
        return count

    def num_of_met_indicator_reports(self, clusters=None, partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_met_indicator_reports(partner=partner)
        return count

    def num_of_constrained_indicator_reports(
            self, clusters=None, partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_constrained_indicator_reports(partner=partner)
        return count

    def num_of_on_track_indicator_reports(self, clusters=None, partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_on_track_indicator_reports(partner=partner)
        return count

    def num_of_no_progress_indicator_reports(
            self, clusters=None, partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_no_progress_indicator_reports(partner=partner)
        return count

    def num_of_no_status_indicator_reports(self, clusters=None, partner=None):
        if not clusters:
            clusters = self.all_clusters

        count = 0
        for c in clusters:
            count += c.num_of_no_status_indicator_reports(partner=partner)
        return count

    def num_of_projects(self, clusters=None, partner=None):
        from partner.models import PartnerProject

        if not clusters:
            clusters = self.all_clusters

        qset = PartnerProject.objects.filter(clusters__in=clusters)
        if partner:
            qset = qset.filter(partner=partner)
        return qset.count()

    def _latest_indicator_reports(self, clusters):
        from indicator.models import Reportable, IndicatorReport
        reportables = Reportable.objects.filter(
            Q(partner_activities__partner__clusters__in=clusters)
        )
        return IndicatorReport.objects.filter(
            reportable__in=reportables).order_by(
            '-time_period_end').distinct()

    def upcoming_indicator_reports(self, clusters=None, partner=None,
                                   limit=None, days=15):
        if not clusters:
            clusters = self.all_clusters

        days_in_future = datetime.today() + timedelta(days=days)
        indicator_reports = self._latest_indicator_reports(clusters).filter(
            report_status=INDICATOR_REPORT_STATUS.due
        ).filter(
            due_date__gte=datetime.today()
        ).filter(
            due_date__lte=days_in_future
        )
        return indicator_reports

    def overdue_indicator_reports(self, clusters=None, partner=None,
                                  limit=None):
        """
        Returns indicator reports associated with partner activities or
        partner projects that are overdue, if partner is specified.
        """
        if not clusters:
            clusters = self.all_clusters

        indicator_reports = self._latest_indicator_reports(clusters)
        indicator_reports = indicator_reports.filter(
            report_status=INDICATOR_REPORT_STATUS.overdue)

        if partner:
            indicator_reports = indicator_reports.filter(
                Q(reportable__partner_projects__partner=partner)
                | Q(reportable__partner_activities__partner=partner)
            )

        if limit:
            indicator_reports = indicator_reports[:limit]

        return indicator_reports

    def constrained_indicator_reports(self, clusters=None, partner=None,
                                      limit=None):
        if not clusters:
            clusters = self.all_clusters

        indicator_reports = self._latest_indicator_reports(clusters)
        if partner:
            indicator_reports = indicator_reports.filter(
                Q(reportable__partner_projects__partner=partner)
                | Q(reportable__partner_activities__partner=partner)
            )
        indicator_reports = indicator_reports.filter(
            report_status=INDICATOR_REPORT_STATUS.accepted,
            overall_status=OVERALL_STATUS.constrained).order_by(
                'reportable__id', '-submission_date'
        ).distinct('reportable__id')
        if limit:
            indicator_reports = indicator_reports[:limit]
        return indicator_reports

    def partner_activities(self, partner, clusters=None, limit=None):
        if not clusters:
            clusters = self.all_clusters
        qset = partner.partner_activities.filter(
            partner__clusters__in=clusters)
        if limit:
            qset = qset[:limit]
        return qset


class GatewayType(TimeStampedModel):
    """
    Represents an Admin Type in location-related models.
    """

    name = models.CharField(max_length=64, unique=True)
    admin_level = models.PositiveSmallIntegerField()

    country = models.ForeignKey(Country, related_name="gateway_types")

    class Meta:
        ordering = ['name']
        verbose_name = 'Location Type'

    def __str__(self):
        return '{} - {}'.format(self.country, self.name)


class LocationManager(models.GeoManager):

    def get_queryset(self):
        return super(LocationManager, self).get_queryset(
        ).select_related('gateway')


@python_2_unicode_compatible
class Location(TimeStampedExternalSourceModel):
    """
    Location model define place where agents are working.
    The background of the location can be:
    Country > Region > City > District/Point.

    Either a point or geospatial object.
    pcode should be unique.

    related models:
        indicator.Reportable (ForeignKey): "reportable"
        core.Location (ForeignKey): "self"
        core.GatewayType: "gateway"
    """
    title = models.CharField(max_length=255)

    gateway = models.ForeignKey(
        GatewayType, verbose_name='Location Type', related_name='locations'
    )
    carto_db_table = models.ForeignKey(
        'core.CartoDBTable',
        related_name="locations",
        blank=True,
        null=True
    )

    latitude = models.DecimalField(
        null=True,
        blank=True,
        max_digits=8,
        decimal_places=5,
        validators=[
            MinValueValidator(Decimal(-90)),
            MaxValueValidator(Decimal(90))
        ]
    )
    longitude = models.DecimalField(
        null=True,
        blank=True,
        max_digits=8,
        decimal_places=5,
        validators=[
            MinValueValidator(Decimal(-180)),
            MaxValueValidator(Decimal(180))
        ]
    )
    p_code = models.CharField(max_length=32, blank=True, null=True, verbose_name='Postal Code')

    parent = models.ForeignKey(
        'self', null=True, blank=True, related_name='children', db_index=True
    )

    geom = models.MultiPolygonField(null=True, blank=True)
    point = models.PointField(null=True, blank=True)
    objects = LocationManager()

    class Meta:
        unique_together = ('title', 'p_code')
        ordering = ['title']

    def __str__(self):
        if self.p_code:
            return '{} ({} {})'.format(
                self.title,
                self.gateway.name,
                "{}: {}".format(
                    'CERD' if self.gateway.name == 'School' else 'PCode', self.p_code or ''
                ))

        return self.title

    @property
    def geo_point(self):
        return self.point if self.point else self.geom.point_on_surface if self.geom else ""

    @property
    def point_lat_long(self):
        return "Lat: {}, Long: {}".format(
            self.point.y,
            self.point.x
        )


class CartoDBTable(MPTTModel):
    """
    Represents a table in CartoDB, it is used to imports locations
    related models:
        core.GatewayType: 'gateway'
        core.Country: 'country'
    """

    domain = models.CharField(max_length=254)
    api_key = models.CharField(max_length=254)
    table_name = models.CharField(max_length=254)
    location_type = models.ForeignKey(GatewayType)
    parent = TreeForeignKey('self', null=True, blank=True,
                            related_name='children', db_index=True)

    country = models.ForeignKey(Country, related_name="carto_db_tables")

    def __str__(self):
        return self.table_name
