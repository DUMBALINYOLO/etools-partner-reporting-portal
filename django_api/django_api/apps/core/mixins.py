from datetime import datetime, timedelta

from django.utils.deconstruct import deconstructible
from django.conf import settings
from django.contrib.auth import get_user_model

from social_core.backends.azuread_b2c import AzureADB2COAuth2
from social_core.pipeline import social_auth
from social_core.pipeline import user as social_core_user

from azure.storage import AccessPolicy, SharedAccessPolicy

from storages.backends.azure_storage import AzureStorage
from storages.utils import setting


def social_details(backend, details, response, *args, **kwargs):
    r = social_auth.social_details(backend, details, response, *args, **kwargs)
    r['details']['idp'] = response.get('idp')

    if not r['details'].get('email'):
        r['details']['email'] = response.get('email')

    return r


def get_username(strategy, details, backend, user=None, *args, **kwargs):
    username = details.get('email')

    try:
        get_user_model().objects.get(username=username)

    except get_user_model().DoesNotExist:
        return
        # return HttpResponseRedirect("/workspace_inactive/")

    return {'username': details.get('email')}


def user_details(strategy, details, user=None, *args, **kwargs):
    # # This is where we update the user
    # # see what the property to map by is here
    # updates_available = False

    if user:
        # user_groups = [group.name for group in user.groups.all()]
        # business_area_code = details.get("business_area_code", 'defaultBA1235')

        # Update username with email and unusable password
        user.username = user.email
        user.first_name = details['first_name']
        user.last_name = details['last_name']
        user.set_unusable_password()
        user.save()

        # TODO: Not using country in profile for PRP at the moment. May need to re-evaluate
        # def update_user_country():
        #     try:
        #         user.profile.country = Country.objects.get(business_area_code=business_area_code)

        #     except Country.DoesNotExist:
        #         user.profile.country = Country.objects.get(name='UAT')

        # if not user.profile.country:
        #     update_user_country()
        #     updates_available = True

        # elif not user.profile.country_override:
        #     # make sure that we update the workspace based business area
        #     if business_area_code != user.profile.country.business_area_code:
        #         update_user_country()
        #         updates_available = True

        # if updates_available:
        #     user.save()
        #     user.profile.save()

    return social_core_user.user_details(strategy, details, user, *args, **kwargs)


class CustomAzureADBBCOAuth2(AzureADB2COAuth2):

    def __init__(self, *args, **kwargs):
        super(CustomAzureADBBCOAuth2, self).__init__(*args, **kwargs)
        self.redirect_uri = settings.FRONTEND_HOST + '/social/complete/azuread-b2c-oauth2/'


@deconstructible
class EToolsAzureStorage(AzureStorage):
    account_name = setting("AZURE_ACCOUNT_NAME")
    account_key = setting("AZURE_ACCOUNT_KEY")
    azure_container = setting("AZURE_CONTAINER")
    azure_ssl = setting("AZURE_SSL")

    auto_sign = setting("AZURE_AUTO_SIGN")
    azure_access_policy_permission = setting("AZURE_ACCESS_POLICY_PERMISSION")
    ap_expiry = setting("AZURE_ACCESS_POLICY_EXPIRY")

    def __init__(self, *args, **kwargs):
        super(EToolsAzureStorage, self).__init__(*args, **kwargs)
        self._connection = None

    def url(self, name):
        if hasattr(self.connection, 'make_blob_url'):
            if self.auto_sign:
                access_policy = AccessPolicy()
                access_policy.start = (datetime.utcnow() + timedelta(seconds=-120)).strftime('%Y-%m-%dT%H:%M:%SZ')
                access_policy.expiry = (
                    datetime.utcnow()
                    + timedelta(seconds=self.ap_expiry)
                ).strftime('%Y-%m-%dT%H:%M:%SZ')
                access_policy.permission = self.azure_access_policy_permission
                sap = SharedAccessPolicy(access_policy)
                sas_token = self.connection.generate_shared_access_signature(
                    self.azure_container,
                    blob_name=name,
                    shared_access_policy=sap,
                )
            else:
                sas_token = None
            return self.connection.make_blob_url(
                container_name=self.azure_container,
                blob_name=name,
                protocol=self.azure_protocol,
                sas_token=sas_token,
            )
        else:
            return "{}{}/{}".format(setting('MEDIA_URL'), self.azure_container, name)
