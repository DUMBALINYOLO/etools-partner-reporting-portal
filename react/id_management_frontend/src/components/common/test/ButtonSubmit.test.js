import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import {Button, CircularProgress} from "@material-ui/core";
import ButtonSubmit from '../ButtonSubmit';

describe('ButtonSubmit component', () => {
    it('renders ButtonSubmit component and calls click properly', () => {
        const classes = {wrapper: 'wrapper', buttonProgress: 'buttonProgress'};
        const loading = false;
        const label = 'Submit';

        const wrapper = shallow(<ButtonSubmit
            className={classes.wrapper}
            disabled={loading}
            buttonLabel={label}
        />);

        wrapper.simulate('submit');

        expect(loading).toBe(false);
        expect(wrapper.dive().length).toBe(1);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders Loading component when loading is true', () => {
        const classes = {wrapper: 'wrapper', buttonProgress: 'buttonProgress'};
        let loading = false;
        const label = false;

        const wrapper = shallow(<ButtonSubmit
            className={classes.wrapper}
            disabled={loading}
            buttonLabel={label}
        />);

        expect(toJSON(wrapper)).toMatchSnapshot();

        loading = true;

        const newWrapper = shallow(<ButtonSubmit
            className={classes.wrapper}
            disabled={loading}
            buttonLabel={label}
        />);

        expect(newWrapper.contains([
            <Button disabled="true">Save</Button>,
            <CircularProgress className={classes.buttonProgress}/>
        ]));

        expect(toJSON(newWrapper)).toMatchSnapshot();
    });
});