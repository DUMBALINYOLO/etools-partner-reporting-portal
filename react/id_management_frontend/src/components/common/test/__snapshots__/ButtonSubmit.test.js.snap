// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSubmit component renders ButtonSubmit component and calls click properly 1`] = `
<WithStyles(ButtonSubmit)
  buttonLabel="Submit"
  classes={
    Object {
      "buttonProgress": "buttonProgress",
      "wrapper": "wrapper",
    }
  }
  loading={false}
>
  <ButtonSubmit
    buttonLabel="Submit"
    classes={
      Object {
        "buttonProgress": "ButtonSubmit-buttonProgress-2 buttonProgress",
        "wrapper": "ButtonSubmit-wrapper-1 wrapper",
      }
    }
    loading={false}
  >
    <div
      className="ButtonSubmit-wrapper-1 wrapper"
    >
      <WithStyles(Button)
        color="primary"
        disabled={false}
        type="submit"
        variant="contained"
      >
        <Button
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit-22",
              "contained": "MuiButton-contained-12",
              "containedPrimary": "MuiButton-containedPrimary-13",
              "containedSecondary": "MuiButton-containedSecondary-14",
              "disabled": "MuiButton-disabled-21",
              "extendedFab": "MuiButton-extendedFab-19",
              "fab": "MuiButton-fab-18",
              "flat": "MuiButton-flat-8",
              "flatPrimary": "MuiButton-flatPrimary-9",
              "flatSecondary": "MuiButton-flatSecondary-10",
              "focusVisible": "MuiButton-focusVisible-20",
              "fullWidth": "MuiButton-fullWidth-26",
              "label": "MuiButton-label-4",
              "mini": "MuiButton-mini-23",
              "outlined": "MuiButton-outlined-11",
              "raised": "MuiButton-raised-15",
              "raisedPrimary": "MuiButton-raisedPrimary-16",
              "raisedSecondary": "MuiButton-raisedSecondary-17",
              "root": "MuiButton-root-3",
              "sizeLarge": "MuiButton-sizeLarge-25",
              "sizeSmall": "MuiButton-sizeSmall-24",
              "text": "MuiButton-text-5",
              "textPrimary": "MuiButton-textPrimary-6",
              "textSecondary": "MuiButton-textSecondary-7",
            }
          }
          color="primary"
          component="button"
          disableFocusRipple={false}
          disabled={false}
          fullWidth={false}
          mini={false}
          size="medium"
          type="submit"
          variant="contained"
        >
          <WithStyles(ButtonBase)
            className="MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
            component="button"
            disabled={false}
            focusRipple={true}
            focusVisibleClassName="MuiButton-focusVisible-20"
            type="submit"
          >
            <ButtonBase
              centerRipple={false}
              className="MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
              classes={
                Object {
                  "disabled": "MuiButtonBase-disabled-28",
                  "focusVisible": "MuiButtonBase-focusVisible-29",
                  "root": "MuiButtonBase-root-27",
                }
              }
              component="button"
              disableRipple={false}
              disableTouchRipple={false}
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="MuiButton-focusVisible-20"
              tabIndex="0"
              type="submit"
            >
              <button
                className="MuiButtonBase-root-27 MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
                disabled={false}
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="submit"
              >
                <span
                  className="MuiButton-label-4"
                >
                  Save
                </span>
                <WithStyles(TouchRipple)
                  center={false}
                  innerRef={[Function]}
                >
                  <TouchRipple
                    center={false}
                    classes={
                      Object {
                        "child": "MuiTouchRipple-child-34",
                        "childLeaving": "MuiTouchRipple-childLeaving-35",
                        "childPulsate": "MuiTouchRipple-childPulsate-36",
                        "ripple": "MuiTouchRipple-ripple-31",
                        "ripplePulsate": "MuiTouchRipple-ripplePulsate-33",
                        "rippleVisible": "MuiTouchRipple-rippleVisible-32",
                        "root": "MuiTouchRipple-root-30",
                      }
                    }
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      className="MuiTouchRipple-root-30"
                      component="span"
                      enter={true}
                      exit={true}
                    >
                      <span
                        className="MuiTouchRipple-root-30"
                      />
                    </TransitionGroup>
                  </TouchRipple>
                </WithStyles(TouchRipple)>
              </button>
            </ButtonBase>
          </WithStyles(ButtonBase)>
        </Button>
      </WithStyles(Button)>
    </div>
  </ButtonSubmit>
</WithStyles(ButtonSubmit)>
`;

exports[`ButtonSubmit component renders Loading component when loading is true 1`] = `
<WithStyles(ButtonSubmit)
  classes={
    Object {
      "buttonProgress": "buttonProgress",
      "wrapper": "wrapper",
    }
  }
  label={null}
  loading={true}
>
  <ButtonSubmit
    classes={
      Object {
        "buttonProgress": "ButtonSubmit-buttonProgress-2 buttonProgress",
        "wrapper": "ButtonSubmit-wrapper-1 wrapper",
      }
    }
    label={null}
    loading={true}
  >
    <div
      className="ButtonSubmit-wrapper-1 wrapper"
    >
      <WithStyles(Button)
        color="primary"
        disabled={true}
        type="submit"
        variant="contained"
      >
        <Button
          classes={
            Object {
              "colorInherit": "MuiButton-colorInherit-22",
              "contained": "MuiButton-contained-12",
              "containedPrimary": "MuiButton-containedPrimary-13",
              "containedSecondary": "MuiButton-containedSecondary-14",
              "disabled": "MuiButton-disabled-21",
              "extendedFab": "MuiButton-extendedFab-19",
              "fab": "MuiButton-fab-18",
              "flat": "MuiButton-flat-8",
              "flatPrimary": "MuiButton-flatPrimary-9",
              "flatSecondary": "MuiButton-flatSecondary-10",
              "focusVisible": "MuiButton-focusVisible-20",
              "fullWidth": "MuiButton-fullWidth-26",
              "label": "MuiButton-label-4",
              "mini": "MuiButton-mini-23",
              "outlined": "MuiButton-outlined-11",
              "raised": "MuiButton-raised-15",
              "raisedPrimary": "MuiButton-raisedPrimary-16",
              "raisedSecondary": "MuiButton-raisedSecondary-17",
              "root": "MuiButton-root-3",
              "sizeLarge": "MuiButton-sizeLarge-25",
              "sizeSmall": "MuiButton-sizeSmall-24",
              "text": "MuiButton-text-5",
              "textPrimary": "MuiButton-textPrimary-6",
              "textSecondary": "MuiButton-textSecondary-7",
            }
          }
          color="primary"
          component="button"
          disableFocusRipple={false}
          disabled={true}
          fullWidth={false}
          mini={false}
          size="medium"
          type="submit"
          variant="contained"
        >
          <WithStyles(ButtonBase)
            className="MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 MuiButton-disabled-21"
            component="button"
            disabled={true}
            focusRipple={true}
            focusVisibleClassName="MuiButton-focusVisible-20"
            type="submit"
          >
            <ButtonBase
              centerRipple={false}
              className="MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 MuiButton-disabled-21"
              classes={
                Object {
                  "disabled": "MuiButtonBase-disabled-28",
                  "focusVisible": "MuiButtonBase-focusVisible-29",
                  "root": "MuiButtonBase-root-27",
                }
              }
              component="button"
              disableRipple={false}
              disableTouchRipple={false}
              disabled={true}
              focusRipple={true}
              focusVisibleClassName="MuiButton-focusVisible-20"
              tabIndex="0"
              type="submit"
            >
              <button
                className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-3 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 MuiButton-disabled-21"
                disabled={true}
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="-1"
                type="submit"
              >
                <span
                  className="MuiButton-label-4"
                >
                  Save
                </span>
              </button>
            </ButtonBase>
          </WithStyles(ButtonBase)>
        </Button>
      </WithStyles(Button)>
      <WithStyles(CircularProgress)
        className="ButtonSubmit-buttonProgress-2 buttonProgress"
        size={24}
      >
        <CircularProgress
          className="ButtonSubmit-buttonProgress-2 buttonProgress"
          classes={
            Object {
              "circle": "MuiCircularProgress-circle-43",
              "circleIndeterminate": "MuiCircularProgress-circleIndeterminate-45",
              "circleStatic": "MuiCircularProgress-circleStatic-44",
              "colorPrimary": "MuiCircularProgress-colorPrimary-40",
              "colorSecondary": "MuiCircularProgress-colorSecondary-41",
              "indeterminate": "MuiCircularProgress-indeterminate-39",
              "root": "MuiCircularProgress-root-37",
              "static": "MuiCircularProgress-static-38",
              "svg": "MuiCircularProgress-svg-42",
            }
          }
          color="primary"
          size={24}
          thickness={3.6}
          value={0}
          variant="indeterminate"
        >
          <div
            className="MuiCircularProgress-root-37 MuiCircularProgress-colorPrimary-40 MuiCircularProgress-indeterminate-39 ButtonSubmit-buttonProgress-2 buttonProgress"
            role="progressbar"
            style={
              Object {
                "height": 24,
                "width": 24,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg-42"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle-43 MuiCircularProgress-circleIndeterminate-45"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </div>
        </CircularProgress>
      </WithStyles(CircularProgress)>
    </div>
  </ButtonSubmit>
</WithStyles(ButtonSubmit)>
`;
